eth_account/_utils/encode_typed_data/encoding_and_hashing.py

```diff
@@ -0,0 +1,209 @@
+from typing import (
+    Any,
+    Dict,
+    List,
+    Tuple,
+    Union,
+)
+
+from eth_abi import (
+    encode,
+)
+from eth_utils import (
+    is_hexstr,
+    keccak,
+    to_bytes,
+    to_int,
+)
+
+from eth_account._utils.encode_typed_data.helpers import (
+    EIP712_SOLIDITY_TYPES,
+    coerce_bool,
+    is_array_type,
+    parse_core_array_type,
+    parse_parent_array_type,
+)
+
+
+def get_primary_type(types: Dict[str, List[Dict[str, str]]]) -> str:
+    custom_types = set(types.keys())
+    custom_types_that_are_deps = set()
+
+    for type_ in custom_types:
+        type_fields = types[type_]
+        for field in type_fields:
+            parsed_type = parse_core_array_type(field["type"])
+            if parsed_type in custom_types and parsed_type != type_:
+                custom_types_that_are_deps.add(parsed_type)
+
+    primary_type = list(custom_types.difference(custom_types_that_are_deps))
+    if len(primary_type) == 1:
+        return primary_type[0]
+    else:
+        raise ValueError("Unable to determine primary type")
+
+
+def encode_field(
+    types: Dict[str, List[Dict[str, str]]],
+    name: str,
+    type_: str,
+    value: Any,
+) -> Tuple[str, Union[int, bytes]]:
+    if type_ in types.keys():
+        # type is a custom type
+        if value is None:
+            return ("bytes32", b"\x00" * 32)
+        else:
+            return ("bytes32", keccak(encode_data(type_, types, value)))
+
+    elif type_ in ["string", "bytes"] and value is None:
+        return ("bytes32", b"")
+
+    # None is allowed only for custom and dynamic types
+    elif value is None:
+        raise ValueError(f"Missing value for field `{name}` of type `{type_}`")
+
+    elif is_array_type(type_):
+        type_ = parse_parent_array_type(type_)
+        type_value_pairs = [encode_field(types, name, type_, item) for item in value]
+        data_types, data_hashes = zip(*type_value_pairs)
+        return ("bytes32", keccak(encode(data_types, data_hashes)))
+
+    elif type_ == "bool":
+        return (type_, coerce_bool(value))
```

Comment:
`bool(value)` seems to be consistent with everything that `eth-sig-util` is expecting. Are there examples where the logic within `coerce_bool()` is expected?

Reply:
Found some cases, actually. For python bytes types, the bool(value) will treat all bytes as "true", whereas javascript / eth-sig-utils will evalutate the value as a number, rendering 0 as False and any other byte value as True. We could likely use if isinstance(value, int) -> bool(to_bytes(value)).

Reply:
Nope, `coerce_bool` was my own "reasonableness" addition. `bool(value)` works for me.